README.md
# lange-notify [![Build Status](https://travis-ci.org/Lange/lange-notify.svg?branch=master)](https://travis-ci.org/Lange/lange-notify)
This is a [NodeCG](http://github.com/nodecg/nodecg) bundle.

This is the notification graphic that I use on [my stream](http://twitch.tv/langeh).
I have open-sourced it so that other developers may learn from it, but I kindly request that it not be used verbatim
on other streams. I worked hard to make something unique that fit well with the rest of my design, and seeing it used
exactly as-is on other streams would be frustrating.

## Demo
<img src="sample.gif?raw=true"/>
(Trimmed and rendered at 30fps to keep GIF size down, actual notification lasts longer and is 60fps)

[60fps YouTube demo w/ sound](https://www.youtube.com/watch?v=yMC4vuoset8)

## Code Walkthrough (v0.0.1, Oct. 6, 2015)
This video goes through every line of code in the first version of bundle. 
It no longer reflects the current state of this bundle, but some may still find it educational. https://www.youtube.com/watch?v=ev0EG4L_Q8s

## Installation
_If you already have NodeCG installed, skip to step #3._

1. Install [Node.js & npm](https://nodejs.org/en/).

2. Open a command prompt and run the following commands to install NodeCG.
Keep this command prompt open after NodeCG has been installed.
	```sh
	npm install -g bower
	npm install -g nodecg-cli
	mkdir nodecg
	cd nodecg
	nodecg setup
	```

3. Install lange-notify and three of the bundles that it depends on.
	```sh
	nodecg install lange/lange-notify
	nodecg install "supportclass/lfg-filter#^3.0.0"
	nodecg install "supportclass/lfg-nucleus#^3.0.0"
	```

4. To listen for subscriptions and resubs, install and configure one of the following:
	- [lfg-sublistener](https://github.com/SupportClass/lfg-sublistener) - **Recommended** Listens directly to Twitch chat.  
		`nodecg install supportclass/lfg-sublistener`
	- [lfg-siphon](https://github.com/SupportClass/lfg-siphon) - Listens to Twitch chat via [Streen](https://github.com/SupportClass/streen).  
		`nodecg install supportclass/lfg-siphon`
  
5. To listen to [StreamTip](https://streamtip.com), install and configure [lfg-streamtip](https://github.com/SupportClass/lfg-streamtip).
	```sh
	nodecg install supportclass/lfg-streamtip
	```

6. Start NodeCG.
	```sh
	nodecg start
	```

### Usage
- Add [`http://localhost:9090/graphics/lange-notify`](http://localhost:9090/graphics/lange-notify) 
(or whatever your NodeCG hostname and port are) as a Browser Source in OBS, with dimensions of 992x100
- Use lfg-nucleus' "Notifications" panel on the dashboard ([`http://localhost:9090`](http://localhost:9090)) to send test subs, resubs, and tips.

### Additional Credits
- Sound design by [Anthony "Airon" Oetzmann](http://aironaudio.weebly.com/) 

### License
lange-notify is provided under the MIT license, which is available to read in the [LICENSE][] file.
[license]: LICENSE


CHANGELOG.md
<a name="3.0.0"></a>
# [3.0.0](https://github.com/Lange/lange-notify/compare/v2.1.0...v3.0.0) (2018-01-16)


### Features

* add support for cheers ([7734aea](https://github.com/Lange/lange-notify/commit/7734aea))
* port to NodeCG v0.9 ([15b579a](https://github.com/Lange/lange-notify/commit/15b579a))


### BREAKING CHANGES

* No longer supports versions of NodeCG earlier than v0.9.0.



<a name="2.1.0"></a>
# [2.1.0](https://github.com/Lange/lange-notify/compare/v2.0.0...v2.1.0) (2017-03-08)


### Features

* add support for lfg-nucleus v3.x.x ([06c579d](https://github.com/Lange/lange-notify/commit/06c579d))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/Lange/lange-notify/compare/v1.0.0...v2.0.0) (2016-06-23)


### Code Refactoring

* port to NodeCG v0.8 ([62d8e06](https://github.com/Lange/lange-notify/commit/62d8e06))


### BREAKING CHANGES

* no longer supports NodeCG v0.7



<a name="1.0.0"></a>
# [1.0.0](https://github.com/Lange/lange-notify/compare/v0.0.1...v1.0.0) (2016-03-02)


### Code Refactoring

* **all:** port to NodeCG v0.7 ([e806674](https://github.com/Lange/lange-notify/commit/e806674))


### BREAKING CHANGES

* all: NodeCG v0.6 is no longer supported





graphics/app/setup.js
/* eslint-disable no-var */
var bgs;
var label;
/* eslint-enable no-var */

(function () {
	'use strict';

	// Create the canvas element that will become the render target.
	// EaselJS calls this a "stage".
	const containerEl = document.getElementById('container');
	const stageEl = document.createElement('canvas');
	stageEl.id = 'notification';
	stageEl.width = 992;
	stageEl.height = 100;
	containerEl.appendChild(stageEl);

	const SLANT = 20;

	// Store some important coordinates we'll be needing later.
	const midX = stageEl.width / 2;
	const maxX = stageEl.width - SLANT;
	const maxY = stageEl.height;

	// Create the stage on the target canvas, and create a ticker that will render at 60 fps.
	const stage = new createjs.Stage('notification');
	createjs.Ticker.addEventListener('tick', event => {
		if (event.paused) {
			return;
		}

		bgs.forEach(bg => {
			// The two left side x values. Will be inverted to get right side x values.
			const tipX = Math.min(-(bg.width / 2), 0);
			const baseX = Math.min(-(bg.width / 2) + SLANT, 0);

			// Start at top left point, moves clockwise
			bg.graphics
				.clear()
				.beginFill(bg.color)
				.moveTo(baseX, 0)
				.lineTo(-baseX, 0)
				.lineTo(-tipX, maxY / 2)
				.lineTo(-baseX, maxY)
				.lineTo(baseX, maxY)
				.lineTo(tipX, maxY / 2)
				.closePath();
		});

		stage.update();
	});
	createjs.Ticker.framerate = 60;
	createjs.Ticker.timingMode = createjs.Ticker.RAF_SYNCHED;

	const bgContainer = new createjs.Container();
	stage.addChild(bgContainer);

	// Create the three background elements
	const bg3 = new createjs.Shape();
	bg3.name = 'bg3';
	bg3.maxWidth = maxX;
	bg3.width = 0;
	bg3.x = midX;
	bgContainer.addChild(bg3);

	const bg2 = new createjs.Shape();
	bg2.name = 'bg2';
	bg2.maxWidth = maxX - 23;
	bg2.width = 0;
	bg2.x = midX;
	bgContainer.addChild(bg2);

	const bg1 = new createjs.Shape();
	bg1.name = 'bg1';
	bg1.maxWidth = maxX - 54;
	bg1.width = 0;
	bg1.x = midX;
	bgContainer.addChild(bg1);

	// Put all the background elements into an array to make the stagger tweens a bit easier to write
	bgs = [bg1, bg2, bg3];

	// Create the text element
	label = new createjs.Text('SUBSCRIBER', '800 65px proxima-nova', '#F7F2E8');
	label.x = midX;
	label.showY = 11;
	label.hideY = 0;
	label.y = label.hideY;
	label.textAlign = 'center';
	label.mask = bg1;
	label.maxWidth = bg1.maxWidth - (SLANT * 3);
	stage.addChild(label);
})();


graphics/app/notify.js
(function () {
	'use strict';

	// Permanent GSAP timeline
	const tl = new TimelineLite({autoRemoveChildren: true});
	const label = window.label;
	const bgs = window.bgs;

	// Dope constants.
	const DELAY_INCREMENT = 0.09;
	const SUB_COLORS = [
		'#e87933',
		'#f6ce14',
		'#6bba82'
	];
	const TIP_COLORS = [
		'#6bba82',
		'#e87933',
		'#f6ce14'
	];
	const FIRST_MSG_FONT = '800 65px proxima-nova';
	const SECOND_MSG_FONT = '800 65px proxima-nova';

	nodecg.listenFor('subscription', 'lfg-nucleus', data => {
		let firstMsg = 'NEW SUBSCRIBER';
		if (data.resub) {
			firstMsg = `RESUB × ${data.months}`;
		}

		notify(firstMsg, data.name, {
			colors: SUB_COLORS,
			inSound: 'sub'
		});
	});

	nodecg.listenFor('tip', 'lfg-nucleus', tip => {
		// Got a tip from StreamTip
		notify(`${tip.formattedAmount} TIP`, truncateTo25(tip.name), {
			colors: TIP_COLORS,
			inSound: 'tip'
		});
	});

	nodecg.listenFor('cheer', 'lfg-nucleus', cheer => {
		// Got a cheer from Twitch
		notify(`${cheer.formattedAmount} BITS`, truncateTo25(cheer.name), {
			colors: TIP_COLORS,
			inSound: 'tip'
		});
	});

	function notify(firstMsg, secondMsg, opts) {
		firstMsg = firstMsg.toUpperCase();
		secondMsg = secondMsg.toUpperCase();
		opts = opts || {};
		opts.colors = opts.colors || SUB_COLORS;

		const reverseBgs = bgs.slice(0).reverse();
		const foremostBg = bgs[0];
		let delay = 0;

		// Animate in
		tl.add('npIn');

		tl.call(() => {
			const len = bgs.length;
			for (let i = 0; i < len; i++) {
				bgs[i].color = opts.colors[i];
			}
			nodecg.playSound(opts.inSound);
		}, null, null, 'npIn');

		reverseBgs.forEach(bg => {
			tl.to(bg, 0.75, {
				width: bg.maxWidth,
				ease: Elastic.easeOut.config(0.3, 0.4)
			}, `npIn+=${delay}`);
			delay += DELAY_INCREMENT;
		});

		tl.to(label, 0.6, {
			onStart() {
				label.font = FIRST_MSG_FONT;
				label.text = firstMsg;
			},
			y: label.showY,
			ease: Back.easeOut.config(4),
			autoRound: false
		}, `npIn+=${delay - DELAY_INCREMENT}`);

		// Show second message
		tl.to(foremostBg, 0.6, {
			onStart() {
				nodecg.playSound('cut');
			},
			width: 0,
			ease: Elastic.easeIn.config(0.3, 0.4),
			onComplete() {
				label.font = SECOND_MSG_FONT;
				label.text = secondMsg;
			}
		}, '+=1.5');

		tl.to(foremostBg, 0.6, {
			width: foremostBg.maxWidth,
			ease: Elastic.easeOut.config(0.3, 0.4)
		}, '+=0.01');

		// Animate out
		delay = 0;

		tl.add('npOut', '+=4');
		tl.call(() => {
			nodecg.playSound('out');
		}, null, null, 'npOut');
		bgs.forEach(bg => {
			tl.to(bg, 0.7, {
				width: 0,
				ease: Elastic.easeIn.config(0.3, 0.4)
			}, `npOut+=${delay}`);
			delay += DELAY_INCREMENT;
		});

		tl.to(label, 0.467, {
			y: label.hideY,
			ease: Power2.easeIn
		}, `npOut+=${delay}`);

		// Kill time between successive notifications
		tl.to({}, 1, {});
	}

	function truncateTo25(text) {
		const len = text.length;
		if (len > 25) {
			let truncated = text.substring(0, 23);
			truncated += '…';
			return truncated;
		}

		return text;
	}
})();
